File: java/src/main/java/com/gigamonkeys/bells/Calendar.java
package com.gigamonkeys.bells;

import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * One academic year's worth of schedule logic. Wraps a {@link CalendarData} together with
 * viewer {@link Options} and answers questions about school days, schedules, and the
 * current interval for instants that fall within this year.
 */
public final class Calendar {

  private static final String[] WEEKDAY_NAMES = {
    "", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"
  };

  private final CalendarData data;
  private final ZoneId timezone;
  private final String role;
  private final Map<Integer, List<String>> includeTags;
  private final LocalDate firstDay;
  private final LocalDate lastDay;
  private final Map<String, List<PeriodData>> schedules;
  private final Map<String, String> weekdaySchedules;
  private final Map<String, DateEntry> dates;
  private final List<String> holidays;
  private final List<String> teacherWorkDays;
  private final Map<String, String> breakNames;
  private final Map<String, String> nonClassDays;

  /**
   * @param data one year's calendar data
   * @param options the viewer options (role + includeTags)
   */
  public Calendar(CalendarData data, Options options) {
    this.data = data;
    this.timezone = ZoneId.of(data.timezone());
    this.role = options.role();
    this.includeTags = options.includeTags();

    boolean teacher = Options.TEACHER.equals(role) && data.firstDayTeachers() != null;
    this.firstDay = DateTimes.parsePlainDate(teacher ? data.firstDayTeachers() : data.firstDay());
    this.lastDay = DateTimes.parsePlainDate(data.lastDay());

    this.schedules = data.schedules();
    this.weekdaySchedules = data.weekdaySchedules();
    this.dates = data.dates();
    this.holidays = data.holidays();
    this.teacherWorkDays = data.teacherWorkDays();
    this.breakNames = data.breakNames();
    this.nonClassDays = data.nonClassDays();
  }

  // ─── Accessors used by Schedule ───────────────────────────────────────────────

  ZoneId timezone() {
    return timezone;
  }

  String role() {
    return role;
  }

  Map<Integer, List<String>> includeTags() {
    return includeTags;
  }

  Map<String, String> breakNames() {
    return breakNames;
  }

  /**
   * @return the first day of school for this calendar (teacher day for the teacher role)
   */
  public LocalDate firstDay() {
    return firstDay;
  }

  /**
   * @return the last day of school for this calendar
   */
  public LocalDate lastDay() {
    return lastDay;
  }

  // ─── Year boundaries ──────────────────────────────────────────────────────────

  /**
   * @param instant a moment
   * @return whether {@code instant} falls within this calendar's school year
   */
  public boolean isInCalendar(Instant instant) {
    return !startOfYear().isAfter(instant) && !instant.isAfter(endOfYear());
  }

  /**
   * @return the instant the school year begins (start of the first day)
   */
  public Instant startOfYear() {
    Schedule sched = schedule(firstDay);
    return sched.firstPeriod().startInstant(firstDay, timezone);
  }

  /**
   * @return the instant the school year ends (end of the last day)
   */
  public Instant endOfYear() {
    Schedule sched = schedule(lastDay);
    return sched.lastPeriod().endInstant(lastDay, timezone);
  }

  // ─── Schedules ────────────────────────────────────────────────────────────────

  /**
   * The schedule in effect on a given date.
   *
   * @param date the date
   * @return the schedule
   */
  public Schedule schedule(LocalDate date) {
    String d = date.toString();
    List<PeriodData> periods;
    String name = null;

    DateEntry entry = dates.get(d);
    if (entry != null && !date.isBefore(firstDay)) {
      if (entry.isName()) {
        periods = namedSchedule(entry.scheduleName());
        name = entry.scheduleName();
      } else {
        periods = entry.periods();
      }
    } else {
      String weekdayName = WEEKDAY_NAMES[date.getDayOfWeek().getValue()];
      name = weekdaySchedules.getOrDefault(weekdayName, "NORMAL");
      periods = namedSchedule(name);
    }

    return new Schedule(this, DateTimes.resolveScheduleTimes(periods), date, name);
  }

  private List<PeriodData> namedSchedule(String name) {
    List<PeriodData> periods = schedules.get(name);
    if (periods == null) {
      throw new IllegalArgumentException("Unknown schedule \"" + name + "\"");
    }
    return periods;
  }

  // ─── School-day predicates ────────────────────────────────────────────────────

  /**
   * @param date a date
   * @return whether it is a school day (weekday and not a holiday)
   */
  public boolean isSchoolDay(LocalDate date) {
    int dow = date.getDayOfWeek().getValue();
    return dow != 6 && dow != 7 && !isHoliday(date);
  }

  /**
   * @param date a date
   * @return whether it is a holiday for this viewer's role
   */
  public boolean isHoliday(LocalDate date) {
    String d = date.toString();
    return holidays.contains(d)
        && !(Options.TEACHER.equals(role) && teacherWorkDays.contains(d));
  }

  /**
   * Find the next holiday after the given instant, or {@code lastDay} if none.
   *
   * @param instant a moment
   * @return the next holiday date
   */
  public LocalDate nextHoliday(Instant instant) {
    LocalDate d = instant.atZone(timezone).toLocalDate().plusDays(1);
    while (!isHoliday(d) && !d.isAfter(lastDay)) {
      d = d.plusDays(1);
    }
    return d;
  }

  // ─── Next/previous school day ─────────────────────────────────────────────────

  private LocalDate nextSchoolDay(LocalDate date) {
    LocalDate d = date.plusDays(1);
    while (!isSchoolDay(d)) {
      d = d.plusDays(1);
    }
    return d;
  }

  private LocalDate previousSchoolDay(LocalDate date) {
    LocalDate d = date.minusDays(1);
    while (!isSchoolDay(d)) {
      d = d.minusDays(1);
    }
    return d;
  }

  /**
   * @param instant a moment
   * @return the instant the next school day starts
   */
  public Instant nextSchoolDayStart(Instant instant) {
    LocalDate date = instant.atZone(timezone).toLocalDate();
    if (isSchoolDay(date)) {
      Instant start = schedule(date).startOfDay(date, timezone);
      if (start.isAfter(instant)) {
        return start;
      }
    }
    LocalDate next = nextSchoolDay(date);
    return schedule(next).startOfDay(next, timezone);
  }

  /**
   * @param instant a moment
   * @return the instant the previous school day ended
   */
  public Instant previousSchoolDayEnd(Instant instant) {
    LocalDate date = instant.atZone(timezone).toLocalDate();
    if (isSchoolDay(date)) {
      Instant end = schedule(date).endOfDay(date, timezone);
      if (end.isBefore(instant)) {
        return end;
      }
    }
    LocalDate prev = previousSchoolDay(date);
    return schedule(prev).endOfDay(prev, timezone);
  }

  // ─── Current interval ─────────────────────────────────────────────────────────

  /**
   * @param instant a moment
   * @return the interval covering it, or {@code null}
   */
  public Interval currentInterval(Instant instant) {
    LocalDate date = instant.atZone(timezone).toLocalDate();
    return schedule(date).currentInterval(instant);
  }

  // ─── Day counting ─────────────────────────────────────────────────────────────

  /**
   * Count school days remaining from {@code instant} through the end of the year.
   *
   * @param instant a moment
   * @return the count
   */
  public int schoolDaysLeft(Instant instant) {
    LocalDate date = instant.atZone(timezone).toLocalDate();
    Schedule sched = schedule(date);
    Instant endOfDay = isSchoolDay(date) ? sched.endOfDay(date, timezone) : null;

    int count = 0;
    if (endOfDay != null && instant.isBefore(endOfDay)) {
      count = 1; // currently a school day, counts as remaining
    }

    // Always start counting from tomorrow regardless.
    LocalDate d = date.plusDays(1);
    while (!d.isAfter(lastDay)) {
      if (isSchoolDay(d)) {
        count++;
      }
      d = d.plusDays(1);
    }
    return count;
  }

  /**
   * @param date a date
   * @return the non-class label for that date, or {@code null}
   */
  public String nonClassLabel(LocalDate date) {
    return nonClassDays.get(date.toString());
  }

  /**
   * Non-class days from {@code instant} through the end of the year, in date order.
   *
   * @param instant a moment
   * @return the matching non-class days
   */
  public List<NonClassDay> nonClassDaysLeft(Instant instant) {
    LocalDate today = instant.atZone(timezone).toLocalDate();
    Schedule todaySched = isSchoolDay(today) ? schedule(today) : null;
    Instant todayEnd = todaySched != null ? todaySched.endOfDay(today, timezone) : null;
    boolean includesToday = todayEnd != null && instant.isBefore(todayEnd);

    List<NonClassDay> result = new ArrayList<>();
    for (Map.Entry<String, String> e : nonClassDays.entrySet()) {
      LocalDate d = DateTimes.parsePlainDate(e.getKey());
      int cmp = d.compareTo(today);
      if (cmp < 0) {
        continue;
      }
      if (cmp == 0 && !includesToday) {
        continue;
      }
      if (d.isAfter(lastDay)) {
        continue;
      }
      result.add(new NonClassDay(d, e.getValue()));
    }
    result.sort((a, b) -> a.date().compareTo(b.date()));
    return result;
  }

  /**
   * Count school days between two dates (inclusive of both endpoints), clamped to this
   * calendar's bounds.
   *
   * @param start the start date
   * @param end the end date
   * @return the count
   */
  public int schoolDaysBetween(LocalDate start, LocalDate end) {
    LocalDate from = start.isBefore(firstDay) ? firstDay : start;
    LocalDate to = end.isAfter(lastDay) ? lastDay : end;

    int count = 0;
    LocalDate d = from;
    while (!d.isAfter(to)) {
      if (isSchoolDay(d)) {
        count++;
      }
      d = d.plusDays(1);
    }
    return count;
  }

  /**
   * Count calendar days remaining until the day after {@code lastDay}.
   *
   * @param instant a moment
   * @return the count
   */
  public int calendarDaysLeft(Instant instant) {
    LocalDate date = instant.atZone(timezone).toLocalDate();
    LocalDate endDate = lastDay.plusDays(1);
    return DateTimes.daysBetween(noonInstant(date), noonInstant(endDate));
  }

  // ─── School time ──────────────────────────────────────────────────────────────

  /**
   * @param instant a moment
   * @return school time remaining from {@code instant} to the end of the year
   */
  public Duration schoolTimeLeft(Instant instant) {
    return schoolTimeBetween(instant, endOfYear());
  }

  /**
   * @param instant a moment
   * @return school time elapsed from the start of the year to {@code instant}
   */
  public Duration schoolTimeDone(Instant instant) {
    return schoolTimeBetween(startOfYear(), instant);
  }

  /**
   * @return the total school time in this year
   */
  public Duration totalSchoolTime() {
    return schoolTimeBetween(startOfYear(), endOfYear());
  }

  /**
   * Total in-session school time between two instants, clamped to this calendar's bounds.
   *
   * @param start the start instant
   * @param end the end instant
   * @return the in-session duration
   */
  public Duration schoolTimeBetween(Instant start, Instant end) {
    // Clamp start/end to calendar bounds.
    Instant calStart = startOfYear();
    Instant calEnd = endOfYear();
    Instant cursor = start.isBefore(calStart) ? calStart : start;
    Instant finish = end.isAfter(calEnd) ? calEnd : end;

    if (!cursor.isBefore(finish)) {
      return Duration.ZERO;
    }

    long totalMillis = 0;
    LocalDate cursorDate = cursor.atZone(timezone).toLocalDate();

    while (!cursorDate.isAfter(lastDay)) {
      // Stop once the start of this day is past the finish time.
      Instant dayMidnight = cursorDate.atStartOfDay(timezone).toInstant();
      if (!dayMidnight.isBefore(finish)) {
        break;
      }

      if (isSchoolDay(cursorDate)) {
        Schedule sched = schedule(cursorDate);
        Instant dayStart = sched.startOfDay(cursorDate, timezone);
        Instant dayEnd = sched.endOfDay(cursorDate, timezone);

        Instant from = cursor.isAfter(dayStart) ? cursor : dayStart;
        Instant to = finish.isBefore(dayEnd) ? finish : dayEnd;

        if (from.isBefore(to)) {
          totalMillis += to.toEpochMilli() - from.toEpochMilli();
        }
      }

      cursorDate = cursorDate.plusDays(1);
    }

    return Duration.ofMillis(totalMillis);
  }

  private Instant noonInstant(LocalDate date) {
    return DateTimes.noon(date).atZone(timezone).toInstant();
  }
}

File: libs/ts/src/calendar.ts
/**
 * Calendar, Schedule, Period, and Interval classes.
 * Uses Temporal throughout. No localStorage, no module-level globals.
 */

import { parsePlainDate, resolveScheduleTimes, daysBetween, noon, includesWeekend } from './datetime.js';
import type {
  IncludeTags,
  IntervalType,
  NonClassDay,
  PeriodData,
  ResolvedPeriod,
  Role,
  YearData,
} from './types.js';

const WEEKDAY_NAMES: Record<number, string> = {
  1: 'monday',
  2: 'tuesday',
  3: 'wednesday',
  4: 'thursday',
  5: 'friday',
  6: 'saturday',
  7: 'sunday',
};

/**
 * Normalize the includeTags option.
 * Accepts either a flat array (same tags every weekday) or a per-day-of-week map.
 * Returns a map: { 1: [...], 2: [...], 3: [...], 4: [...], 5: [...] }
 */
const normalizeIncludeTags = (includeTags: IncludeTags): Record<number, string[]> => {
  if (Array.isArray(includeTags)) {
    return { 1: includeTags, 2: includeTags, 3: includeTags, 4: includeTags, 5: includeTags };
  }
  return includeTags || {};
};

interface CalendarOptions {
  role: Role;
  includeTags: IncludeTags;
}

class Calendar {
  data: YearData;
  timezone: string;
  role: Role;
  includeTags: Record<number, string[]>;
  firstDay: Temporal.PlainDate;
  lastDay: Temporal.PlainDate;
  schedules: Record<string, PeriodData[]>;
  weekdaySchedules: Record<string, string>;
  dates: Record<string, string | PeriodData[]>;
  holidays: string[];
  teacherWorkDays: string[];
  breakNames: Record<string, string>;
  nonClassDays: Record<string, string>;

  #noonInstant(date: Temporal.PlainDate): Temporal.Instant {
    return noon(date).toZonedDateTime(this.timezone).toInstant();
  }

  #nextSchoolDay(date: Temporal.PlainDate): Temporal.PlainDate {
    let d = date.add({ days: 1 });
    while (!this.isSchoolDay(d)) {
      d = d.add({ days: 1 });
    }
    return d;
  }

  #previousSchoolDay(date: Temporal.PlainDate): Temporal.PlainDate {
    let d = date.subtract({ days: 1 });
    while (!this.isSchoolDay(d)) {
      d = d.subtract({ days: 1 });
    }
    return d;
  }

  #schoolTimeBetween(start: Temporal.Instant, end: Temporal.Instant): Temporal.Duration {
    // Clamp start/end to calendar bounds.
    const calStart = this.startOfYear();
    const calEnd = this.endOfYear();
    const cursor = Temporal.Instant.compare(start, calStart) < 0 ? calStart : start;
    const finish = Temporal.Instant.compare(end, calEnd) > 0 ? calEnd : end;

    if (Temporal.Instant.compare(cursor, finish) >= 0) {
      return Temporal.Duration.from({ seconds: 0 });
    }

    let totalMillis = 0;
    let cursorDate = cursor.toZonedDateTimeISO(this.timezone).toPlainDate();

    while (Temporal.PlainDate.compare(cursorDate, this.lastDay) <= 0) {
      // Stop once the start of this day is past the finish time.
      const dayMidnightInstant = cursorDate.toPlainDateTime({ hour: 0 })
        .toZonedDateTime(this.timezone).toInstant();
      if (Temporal.Instant.compare(dayMidnightInstant, finish) >= 0) break;

      if (this.isSchoolDay(cursorDate)) {
        const sched = this.schedule(cursorDate);
        const dayStart = sched.startOfDay(cursorDate, this.timezone);
        const dayEnd = sched.endOfDay(cursorDate, this.timezone);

        const from = Temporal.Instant.compare(cursor, dayStart) > 0 ? cursor : dayStart;
        const to = Temporal.Instant.compare(finish, dayEnd) < 0 ? finish : dayEnd;

        if (Temporal.Instant.compare(from, to) < 0) {
          totalMillis += to.epochMilliseconds - from.epochMilliseconds;
        }
      }

      cursorDate = cursorDate.add({ days: 1 });
    }

    return Temporal.Duration.from({ milliseconds: totalMillis });
  }

  constructor(data: YearData, options: CalendarOptions) {
    this.data = data;
    this.timezone = data.timezone;
    this.role = options.role;
    this.includeTags = normalizeIncludeTags(options.includeTags);

    this.firstDay = parsePlainDate(
      this.role === 'teacher' && data.firstDayTeachers ? data.firstDayTeachers : data.firstDay
    );
    this.lastDay = parsePlainDate(data.lastDay);
    this.schedules = data.schedules;
    this.weekdaySchedules = data.weekdaySchedules || {};
    this.dates = data.dates || {};
    this.holidays = data.holidays || [];
    this.teacherWorkDays = data.teacherWorkDays || [];
    this.breakNames = data.breakNames || {};
    this.nonClassDays = data.nonClassDays || {};
  }

  isInCalendar(instant: Temporal.Instant): boolean {
    return (
      Temporal.Instant.compare(this.startOfYear(), instant) <= 0 &&
      Temporal.Instant.compare(instant, this.endOfYear()) <= 0
    );
  }

  startOfYear(): Temporal.Instant {
    const sched = this.schedule(this.firstDay);
    return sched.firstPeriod().startInstant(this.firstDay, this.timezone);
  }

  endOfYear(): Temporal.Instant {
    const sched = this.schedule(this.lastDay);
    return sched.lastPeriod().endInstant(this.lastDay, this.timezone);
  }

  schedule(date: Temporal.PlainDate): Schedule {
    const d = date.toString();
    let periods: PeriodData[];
    let name: string | null = null;
    const entry = this.dates[d];
    if (entry !== undefined && d >= this.firstDay.toString()) {
      if (typeof entry === 'string') {
        periods = this.#namedSchedule(entry);
        name = entry;
      } else {
        periods = entry;
      }
    } else {
      const weekdayName = WEEKDAY_NAMES[date.dayOfWeek];
      name = this.weekdaySchedules[weekdayName] || 'NORMAL';
      periods = this.#namedSchedule(name);
    }
    return new Schedule(this, resolveScheduleTimes(periods), date, name);
  }

  #namedSchedule(name: string): PeriodData[] {
    const periods = this.schedules[name];
    if (!periods) {
      throw new Error(`Unknown schedule "${name}"`);
    }
    return periods;
  }

  isSchoolDay(date: Temporal.PlainDate): boolean {
    const dow = date.dayOfWeek;
    return dow !== 6 && dow !== 7 && !this.isHoliday(date);
  }

  isHoliday(date: Temporal.PlainDate): boolean {
    const d = date.toString();
    return (
      this.holidays.includes(d) &&
      !(this.role === 'teacher' && this.teacherWorkDays.includes(d))
    );
  }

  /** Find the next holiday after the given instant. */
  nextHoliday(instant: Temporal.Instant): Temporal.PlainDate {
    let d = instant.toZonedDateTimeISO(this.timezone).toPlainDate().add({ days: 1 });
    while (!this.isHoliday(d) && Temporal.PlainDate.compare(d, this.lastDay) <= 0) {
      d = d.add({ days: 1 });
    }
    return d;
  }

  nextSchoolDayStart(instant: Temporal.Instant): Temporal.Instant {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    if (this.isSchoolDay(date)) {
      const start = this.schedule(date).startOfDay(date, this.timezone);
      if (Temporal.Instant.compare(start, instant) > 0) {
        return start;
      }
    }
    const next = this.#nextSchoolDay(date);
    return this.schedule(next).startOfDay(next, this.timezone);
  }

  previousSchoolDayEnd(instant: Temporal.Instant): Temporal.Instant {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    if (this.isSchoolDay(date)) {
      const end = this.schedule(date).endOfDay(date, this.timezone);
      if (Temporal.Instant.compare(end, instant) < 0) {
        return end;
      }
    }
    const prev = this.#previousSchoolDay(date);
    return this.schedule(prev).endOfDay(prev, this.timezone);
  }

  currentInterval(instant: Temporal.Instant): Interval | null {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const sched = this.schedule(date);
    return sched.currentInterval(instant);
  }

  /** Count school days remaining from `instant` through end of year. */
  schoolDaysLeft(instant: Temporal.Instant): number {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const sched = this.schedule(date);
    const endOfDay = this.isSchoolDay(date) ? sched.endOfDay(date, this.timezone) : null;

    let count = 0;
    if (endOfDay && Temporal.Instant.compare(instant, endOfDay) < 0) {
      count = 1; // currently a school day, counts as remaining
    }

    // Always start counting from tomorrow regardless.
    let d = date.add({ days: 1 });
    const endDate = this.lastDay;

    while (Temporal.PlainDate.compare(d, endDate) <= 0) {
      if (this.isSchoolDay(d)) count++;
      d = d.add({ days: 1 });
    }
    return count;
  }

  nonClassLabel(date: Temporal.PlainDate): string | null {
    return this.nonClassDays[date.toString()] || null;
  }

  /** Non-class days from `instant` through end of year, in date order. */
  nonClassDaysLeft(instant: Temporal.Instant): NonClassDay[] {
    const today = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const todaySched = this.isSchoolDay(today) ? this.schedule(today) : null;
    const todayEnd = todaySched ? todaySched.endOfDay(today, this.timezone) : null;
    const includesToday = todayEnd && Temporal.Instant.compare(instant, todayEnd) < 0;

    const result: NonClassDay[] = [];
    for (const [dateStr, label] of Object.entries(this.nonClassDays)) {
      const d = parsePlainDate(dateStr);
      const cmp = Temporal.PlainDate.compare(d, today);
      if (cmp < 0) continue;
      if (cmp === 0 && !includesToday) continue;
      if (Temporal.PlainDate.compare(d, this.lastDay) > 0) continue;
      result.push({ date: d, label });
    }
    result.sort((a, b) => Temporal.PlainDate.compare(a.date, b.date));
    return result;
  }

  /** Count school days between two plain dates (inclusive of both endpoints). */
  schoolDaysBetween(start: Temporal.PlainDate, end: Temporal.PlainDate): number {
    // Clamp to calendar bounds.
    const from = Temporal.PlainDate.compare(start, this.firstDay) < 0 ? this.firstDay : start;
    const to = Temporal.PlainDate.compare(end, this.lastDay) > 0 ? this.lastDay : end;

    let count = 0;
    let d = from;
    while (Temporal.PlainDate.compare(d, to) <= 0) {
      if (this.isSchoolDay(d)) count++;
      d = d.add({ days: 1 });
    }
    return count;
  }

  /** Count calendar days remaining until the day after lastDay. */
  calendarDaysLeft(instant: Temporal.Instant): number {
    const date = instant.toZonedDateTimeISO(this.timezone).toPlainDate();
    const endDate = this.lastDay.add({ days: 1 });
    return daysBetween(this.#noonInstant(date), this.#noonInstant(endDate));
  }

  /** Compute total school time (as Duration) from `instant` to end of year. */
  schoolTimeLeft(instant: Temporal.Instant): Temporal.Duration {
    return this.#schoolTimeBetween(instant, this.endOfYear());
  }

  /** Compute school time (as Duration) from start of year to `instant`. */
  schoolTimeDone(instant: Temporal.Instant): Temporal.Duration {
    return this.#schoolTimeBetween(this.startOfYear(), instant);
  }

  /** Total school time in the year. */
  totalSchoolTime(): Temporal.Duration {
    return this.#schoolTimeBetween(this.startOfYear(), this.endOfYear());
  }

  /**
   * Public entry point for cross-instance school-time computation.
   * Clamps to this calendar's bounds before summing.
   */
  schoolTimeBetween(start: Temporal.Instant, end: Temporal.Instant): Temporal.Duration {
    return this.#schoolTimeBetween(start, end);
  }
}

class Schedule {
  calendar: Calendar;
  date: Temporal.PlainDate;
  name: string | null;
  rawPeriods: Period[];

  constructor(
    calendar: Calendar,
    periods: ResolvedPeriod[],
    date: Temporal.PlainDate,
    name: string | null = null,
  ) {
    this.calendar = calendar;
    this.date = date;
    this.name = name;
    this.rawPeriods = periods.map((x) => new Period(x.name, x.start, x.end, x.tags, x.teachers));

    // Set .next links on actual periods.
    const actual = this.actualPeriods();
    actual.forEach((p, i) => {
      p.next = i < actual.length - 1 ? actual[i + 1] : null;
    });
  }

  #maybeBreak(instant: Temporal.Instant): Interval | null {
    if (this.notInSchool(instant)) {
      const prev = this.calendar.previousSchoolDayEnd(instant);
      const next = this.calendar.nextSchoolDayStart(instant);
      const days = daysBetween(prev, next);
      if (days >= 3) {
        const name = this.#breakName(days, prev, next);
        return new Interval(`${name}!`, prev, next, false, 'break', []);
      }
    }
    return null;
  }

  #breakName(days: number, start: Temporal.Instant, end: Temporal.Instant): string {
    const tz = this.calendar.timezone;
    if (days > 4) {
      const nextHoliday = this.calendar.nextHoliday(start);
      return this.calendar.breakNames[nextHoliday.toString()] || 'Vacation';
    } else if (includesWeekend(start, end, tz)) {
      return days > 3 ? 'Long weekend' : 'Weekend';
    } else {
      return 'Mid-week vacation?';
    }
  }

  /** Determine if a period should be included given the current date/config. */
  hasPeriod(p: Period): boolean {
    if (p.teachers) {
      return this.calendar.role === 'teacher';
    }

    const tags = p.tags || [];
    if (!tags.includes('optional')) {
      // Not optional — always include.
      return true;
    }

    // Optional — include only if one of the other tags appears in includeTags for this day.
    const dow = this.date.dayOfWeek; // 1=Mon…7=Sun
    const allowed = this.calendar.includeTags[dow] || [];
    return tags.some((tag) => tag !== 'optional' && allowed.includes(tag));
  }

  actualPeriods(): Period[] {
    const base = this.rawPeriods.filter((p) => this.hasPeriod(p));

    if (base.length === 0) return base;

    // Trim nonschool optional periods from start and end. These are administrative
    // periods (e.g. Food Trucks) that should not define school day boundaries.
    // User-configurable optional periods (zero, seventh, ext) are kept so that
    // enabling them correctly affects the start/end of the school day.
    const isNonschool = (p: Period) => (p.tags || []).includes('nonschool');
    while (base.length > 0 && (base[0].tags || []).includes('optional') && isNonschool(base[0])) base.shift();
    while (base.length > 0 && (base[base.length - 1].tags || []).includes('optional') && isNonschool(base[base.length - 1])) base.pop();

    return base;
  }

  firstPeriod(): Period {
    return this.actualPeriods()[0];
  }

  lastPeriod(): Period {
    const ps = this.actualPeriods();
    return ps[ps.length - 1];
  }

  startOfDay(date: Temporal.PlainDate, timezone: string): Temporal.Instant {
    return this.firstPeriod().startInstant(date, timezone);
  }

  endOfDay(date: Temporal.PlainDate, timezone: string): Temporal.Instant {
    return this.lastPeriod().endInstant(date, timezone);
  }

  notInSchool(instant: Temporal.Instant): boolean {
    const date = this.date;
    const tz = this.calendar.timezone;
    return (
      !this.calendar.isSchoolDay(date) ||
      Temporal.Instant.compare(instant, this.endOfDay(date, tz)) >= 0 ||
      Temporal.Instant.compare(instant, this.startOfDay(date, tz)) <= 0
    );
  }

  currentInterval(instant: Temporal.Instant): Interval | null {
    const daysOff = this.#maybeBreak(instant);
    if (daysOff) return daysOff;

    const tz = this.calendar.timezone;
    const date = this.date;
    const first = this.firstPeriod();
    const last = this.lastPeriod();

    if (!first) return null;

    if (first.isAfter(instant, date, tz)) {
      return new Interval(
        'Before school',
        this.calendar.previousSchoolDayEnd(instant),
        first.startInstant(date, tz),
        false,
        'before-school',
        [],
      );
    } else if (last.isBefore(instant, date, tz)) {
      return new Interval(
        'After school',
        last.endInstant(date, tz),
        this.calendar.nextSchoolDayStart(instant),
        false,
        'after-school',
        [],
      );
    } else {
      for (let p: Period | null = first; p !== null; p = p.next) {
        if (p.contains(instant, date, tz)) {
          return p.toInterval(date, tz);
        } else if (p.next && p.isBefore(instant, date, tz) && p.next.isAfter(instant, date, tz)) {
          return new Interval(
            `Passing to ${p.next.name}`,
            p.endInstant(date, tz),
            p.next.startInstant(date, tz),
            true,
            'passing',
            [],
          );
        }
      }
    }

    return null;
  }
}

class Period {
  name: string;
  start: Temporal.PlainTime;
  end: Temporal.PlainTime;
  tags: string[];
  teachers: boolean;
  next: Period | null;

  constructor(
    name: string,
    start: Temporal.PlainTime,
    end: Temporal.PlainTime,
    tags: string[] | undefined,
    teachers: boolean | undefined,
  ) {
    this.name = name;
    this.start = start;
    this.end = end;
    this.tags = tags || [];
    this.teachers = !!teachers;
    this.next = null;
  }

  startInstant(date: Temporal.PlainDate, timezone: string): Temporal.Instant {
    return date.toPlainDateTime(this.start).toZonedDateTime(timezone).toInstant();
  }

  endInstant(date: Temporal.PlainDate, timezone: string): Temporal.Instant {
    return date.toPlainDateTime(this.end).toZonedDateTime(timezone).toInstant();
  }

  // Periods are half-open intervals [start, end): a period owns its start
  // instant but not its end (the end belongs to the following passing period,
  // break, or after-school span). This keeps every boundary instant in exactly
  // one interval rather than briefly falling into none.

  isAfter(instant: Temporal.Instant, date: Temporal.PlainDate, timezone: string): boolean {
    return Temporal.Instant.compare(this.startInstant(date, timezone), instant) > 0;
  }

  isBefore(instant: Temporal.Instant, date: Temporal.PlainDate, timezone: string): boolean {
    return Temporal.Instant.compare(this.endInstant(date, timezone), instant) <= 0;
  }

  contains(instant: Temporal.Instant, date: Temporal.PlainDate, timezone: string): boolean {
    return (
      Temporal.Instant.compare(this.startInstant(date, timezone), instant) <= 0 &&
      Temporal.Instant.compare(instant, this.endInstant(date, timezone)) < 0
    );
  }

  toInterval(date: Temporal.PlainDate, timezone: string): Interval {
    return new Interval(
      this.name,
      this.startInstant(date, timezone),
      this.endInstant(date, timezone),
      true,
      'period',
      this.tags,
    );
  }
}

class Interval {
  name: string;
  start: Temporal.Instant;
  end: Temporal.Instant;
  duringSchool: boolean;
  type: IntervalType;
  tags: string[];

  constructor(
    name: string,
    start: Temporal.Instant,
    end: Temporal.Instant,
    duringSchool: boolean,
    type: IntervalType,
    tags: string[],
  ) {
    this.name = name;
    this.start = start;
    this.end = end;
    this.duringSchool = duringSchool;
    this.type = type;
    this.tags = tags;
  }

  left(now: Temporal.Instant = Temporal.Now.instant()): Temporal.Duration {
    return now.until(this.end);
  }

  done(now: Temporal.Instant = Temporal.Now.instant()): Temporal.Duration {
    return this.start.until(now);
  }
}

export { Calendar, Schedule, Period, Interval, normalizeIncludeTags };
export type { CalendarOptions };
